### DISCUSSION TITLE
TypiedDict and dict compatibility

### DISCUSSION THREAD

**Discussion by mgmira** [microsoft/pyright#5367] [Q&A] [answered]
In the following code, Pyright reports an error because TypedDict X is not considered as a valid dict[str, Any]. I guess this is not a bug, so what is the rational behind this behavior ?
I've looked in the different issue, but couldn't find an answer. Sorry if this has already been discussed.

```python
# pyright: strict

from __future__ import annotations

from typing import Any, TypedDict


class X(TypedDict):
    x: int


def f(x: dict[str, Any]):
    ...


def g(x: X):
    f(x)  # Argument of type "X" cannot be assigned to parameter "x" of type "dict[str, Any]" in function "f"
```

**Comment by hmc-cs-mdrissi** [comment, marked as answer]
dict are Mutable. TypedDict only have string keys. So if you had,

```python
def f(x: dict[str, Any]):
  x[1] = "10"
```

that would be type unsafe to pass a TypedDict.

I think,

```python
def f(x: Mapping[str, Any]):
  f(x)
```

should be type safe and allowed.

**Reply by erictraut** [reply]
For a more complete discussion, you might find the information in [PEP 589](https://peps.python.org/pep-0589/#type-consistency) useful. This is the PEP that introduced `TypedDict` to the Python type system.
